### ISSUE TITLE
Refactor the tests to use the standard Perl testing frameworks

### ISSUE THREAD

**Issue by szabgab** [dk/Prima#5]
Test::More
and to have the test files in the t/ directory

**Comment by Sweet-kid** [comment]
Hi @dk 
I've tried to port one test Object/Application.t to Test::More : https://github.com/Sweet-kid/Prima/commit/04c1fd7314ccdc9a05e9ce88da9fe23ef6862f42 
If the changes make sense to you, then I will go ahead with porting other tests & get someone to test it on Windows as well.

Thanks.

**Comment by dk** [comment]
Hi,

Yes, that makes sense. I'd prefer a module (something like Prima::Test) instead of "do ../Tester.pl" ,
and explicitly declared number of tests, but otherwise I think that actually might work. But one
important note: this test should return a sensible result in no-X11 environment (SKIP or something)

/dk

On Sun, Feb 22, 2015 at 01:33:51PM -0800, Upasana wrote:
> Hi @dk 
> I've tried to port one test Object/Application.t to Test::More : https://github.com/Sweet-kid/Prima/commit/04c1fd7314ccdc9a05e9ce88da9fe23ef6862f42 
> If the changes make sense to you, then I will go ahead with porting other tests & get someone to test it on Windows as well.
> 
> Thanks.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dk/Prima/issues/5#issuecomment-75461455

-- 
Sincerely,
	Dmitry Karasik

**Comment by Sweet-kid** [comment]
Thanks for the comment.
Yep, having Prima::Test module makes more sense.
Sometimes, declaring number of tests explicitly is not always necessary afaik, but I need to study more about when it's necessary.
I will try to skip the tests if no-X11 is true.

If I'll have any questions, then I will ask you. Thanks so much for all the help.

**Comment by Sweet-kid** [comment]
@dk Hi! I've some questions :
1. Why're you especially checking for Windows here https://github.com/dk/Prima/blob/master/test/Tester.pl#L279 ? I searched a bit, but couldn't find a good answer.
2. What's the purpose of __end_wait, wait and __dong exactly? Sorry, but I don't understand it.
3. As per my understanding, Tester.pl checks for X11 and runs tests. If noX11 then it skips tests with warnings. Am I missing something?

Thanks.

**Comment by dk** [comment]
Hi!
1. That's because Tester.pl catches output of individual tests; you don't need that because each test will become a standalone program and Test::More will do that for you.
2. These three come into an interplay, the reason is to test that an event we trigger indeed is fired. __wait is the base call that waits for an event or a timeout. The test can quickly abort the waiting loop by calling __end_wait or __dong, meaning that yes indeed, we've got the expected event, all is fine. __end_wait is used only once and I think should be killed. Also, $dong is rather lame :) If you could change the waiting API with proper functions like begin_loop/end_loop/wait_for_event/event_status or anything in this direction it would be much more readable.
3. Yes, you're right. Now each test has to do the x11 check by itself, but instead of warning it would call skip_all and quit if there's no x11.

Hope this helps
/dk

**Comment by dk** [comment]
update: for speed x11 checking you could possibly make a special test that runs first in the suite, that creates/removes a test/x11-disabled file that Prima::Test would recognize and avoid a potentially long x11 startup call.

**Comment by Sweet-kid** [comment]
Thanks for answering my questions, it helped.
One more : I've not looked into all the test cases, I've checked only Object/*, are there any test cases which run when noX11?

**Comment by dk** [comment]
Yes, here how's it looks under no-x11 run:

dk@udayin src/Prima> make test
PERL_DL_NONLAZY=1 /usr/home/dk/perl5/perlbrew/perls/perl-5.18.2/bin/perl "-Iblib/lib" "-Iblib/arch" test.pl
No protocol specified
** Warning: skipping X11 tests
Testing ./test/Object/Object...passed
Testing ./test/Image/Codec...passed
Testing ./test/Image/Stream...passed
Testing ./test/Image/Colormap...passed
Testing ./test/Image/Import...passed
Total tests passed:5, skipped:25, failed:0
All tests successful

**Comment by Sweet-kid** [comment]
Alright, thanks so much for all the help! :)

**Comment by dk** [comment]
you're welcome!

**Comment by Sweet-kid** [comment]
@dk I've submitted a PR which is moving all tests to t/ https://github.com/dk/Prima/pull/27
When you will merge this PR, then I will send another PR which is porting tests to Test::More. I've managed to port most of the tests (in Object/ & Image/), but not all, I will get back to you with more detailed description of the problem.

Sending two PRs for simplifying the review process.

Thanks.

**Comment by dk** [comment]
Hi Upasana,

I don't like merging a half-baked commit, which leaves the project in an unusable state. Could you continue the work on the forked branch until it all works? Then I think it would be a good time to merge.

/dk

**Comment by Sweet-kid** [comment]
Hi @dk 

Alright, your concern is justified! :)

I've ported those tests which are straight forward. Those which I've not yet ported are not that straight forward & need some effort. Could you please look at t/lib/Prima/Test.pm and confirm if I'm doing everything right?

Thanks.

**Comment by dk** [comment]
It seems to be moving in the right direction, yes. Just it has to be in Prima/Test.pm, not t/lib/Prima/Test.pm . I'd like to see a nicer API for the module, that is exported by default,
where the x11 option could be mentioned in the "use" statement, so that all normal
tests would just say "use Prima::Test;" which would exit gracefully under no-x11 environment.
The tests that are ok with no-x11 env should say something like "use Prima::Test qw(:noX11);".
Thereafter, these can use wait_for_event (instead of __wait), and stop_waiting (instead of end_wait and writing to $dong), and event_status (instead of checking $dong). Something like this:

#!perl
use strict;
use warnings;
use Test::More tests => 1;
use Prima::Test;
ok( defined $::application );

but that's not necessarily plays nice in no-x11 env, so some experimentation is needed

Hope this helps!
/dk

**Comment by Sweet-kid** [comment]
Thanks for the explanation, it helped. Sorry, I misunderstood before, but now I've understood, earlier I made Prima::Test as part of the test suite & not the API for the module. 
Will get back to you soon!

**Comment by Sweet-kid** [comment]
Hi @dk 

I've pushed some commits https://github.com/dk/Prima/pull/27 
Could you please review them?

Now, I am stuck with only one test, t/Object/Component.t
Here are the changes which I made & the output which I get  https://gist.github.com/Sweet-kid/a29ebe00cf0a71c4de18

Thanks.

**Comment by dk** [comment]
Hi there,

Yes it looks definitely better! I have some notes about the smaller stuff and some notes about API.
First the small things:

- we need use strict/use warnings everywhere
- "cmp_ok( $a, '==', $b)" is better off with "is($a,$b,'test name')"
- some fail(0) typos, meant to be fail('test name')
- set_dong should be called as set_flag() or set_flag(0) instead, and exported by default along with other methods
- end_wait should be called as set_flag(1) instead
- $w should not be created at start, but an exported by default function create_window should return it
("my $w = create_window")
- reading $Prima::Test::dong should be get_flag(), also exported by default
- $Prima::Test::dong = 0 should be reset_flag instead
- Prima::Test should not export any variables, @ISA etc ( and not declare them unless they are used )
- I don't like done_testing, I prefer explicitly declaring number of tests at the start . This requires some attention f.ex. here https://github.com/Sweet-kid/Prima/blob/32e7aa0039dcf77b053d82ccc684a9ce4e3dad92/t/Widget/leshed.t#L49 where you'll need skip($why,3) instead of just skip
- tabs alignment sometimes gets applied badly (f.ex. https://github.com/Sweet-kid/Prima/blob/32e7aa0039dcf77b053d82ccc684a9ce4e3dad92/t/Widget/leshed.t#L49 )
- MANIFEST needs to be updated with new test names and removed files

Now with the larger stuff. I'd think that only tests that are no-x11 minority should call "use Prima::Test qw(:noX11);", others should just go with "use Prima::Test". This makes "plan skip_all => 'Skipping all because noX11'" go away (i.e. Prima::Test deals with it itself), but you'll have to experiment whether that would work nicely with Test::More and skip_all.
 
Great work!
Hope this helps
/dk

**Comment by dk** [comment]
update, I've just run t/Object/Component.t from https://gist.github.com/Sweet-kid/a29ebe00cf0a71c4de18 runs fine, no failures

**Comment by Sweet-kid** [comment]
Hi @dk 

Thanks for the review.

I'm working on the changes, will get back to you in a day or two. For now, I've some questions:

* set_dong should be called as set_flag() or set_flag(0) instead, and exported by default along with other methods
* end_wait should be called as set_flag(1) instead
* reading $Prima::Test::dong should be get_flag(), also exported by default
* $Prima::Test::dong = 0 should be reset_flag instead

This sounds fine, but I'm stuck because at some places we need to pass a CODE ref, e.g. this diff:

-my $c = $Prima::Test::w-> insert( Widget =>
-       onCreate  => \&Prima::Test::set_dong,
-       onDestroy => \&Prima::Test::set_dong,
-       onPostMessage => sub { $Prima::Test::dong = 1; @xpm = ($_[1],$_[2])}
+my $c = $window-> insert( Widget =>
+       onCreate  => \&Prima::Test::set_flag,
+       onDestroy => \&Prima::Test::set_flag,
+       onPostMessage => sub { set_flag(0); @xpm = ($_[1],$_[2])}
 );

onCreate & onDestroy expect CODE ref
set_dong (with no arguments) sets dong=1, but set_flag(0) sets dong=1, so I'm not getting how to pass the CODE ref with an argument, so that set_dong will have the same effect as set_flag(0). I searched a bit, but didn't find how to do it. Or should I change the behaviour of set_flag() to be the same as set_flag(0)? If you've any other way to solve it, then please let me know.

**Comment by dk** [comment]
Yes, you're correct - set_flag without arguments should be equivalent to set_flag(0). That way, you can say onCreate => \&set_flag.

**Comment by Sweet-kid** [comment]
Hi @dk 

Again, some questions:

1) Which tests these two skip are skipping exactly?
    https://github.com/Sweet-kid/Prima/blob/master/test/Object/Unicode.t#L7
    https://github.com/Sweet-kid/Prima/blob/master/test/Image/Codec.t#L35

2) Three tests are failing : https://gist.github.com/Sweet-kid/5125abf9201d3b012b4e
     This surprises me most, I don't know what I am doing wrong here https://gist.github.com/Sweet-kid/5125abf9201d3b012b4e#file-timer-t-L17

3) For this :
Now with the larger stuff. I'd think that only tests that are no-x11 minority should call "use Prima::Test qw(:noX11);", others should just go with "use Prima::Test". This makes "plan skip_all => 'Skipping all because noX11'" go away (i.e. Prima::Test deals with it itself), but you'll have to experiment whether that would work nicely with Test::More and skip_all.

Making noX11 a function and then calling it from all the noX11 tests, does it make sense?

**Comment by Sweet-kid** [comment]
For the Component.t failure, I'm running them on linux, which is your OS? Can it be due to OS?

Thanks.

**Comment by dk** [comment]
Hi Upasana,

1) Unicode.t sets the test environment as if there's just one test, named "support", and skips it. For the rewrite we want a skip_all. The same is with Codec.t - it fakes the test env as if it had 3 tests, passes 2 of them, skips the 3rd. Again, the solution is to skip the rest of the tests.

2) This is partially due to me not thinking that set_flag could be called from object context. This patch to set_flag() will help:

```perl
 sub set_flag {
     shift if ref $_[0]; # ok to call as a method
     my $flag = shift;
```

Timer.t fails because there's a missing "insert" on line 15.

3) Sorry i wasn't clear about what is :noX11. This is an import pragma, which should be explicitly tested for in Prima::Test::import() . It's also a bit tricky because it needs to call the Exporter::import(), to have it successfully export set_flag etc names mentioned in @EXPORT. It should work like this:

```perl
use base qw(Exporter);
our @EXPORT = qw(create_window set_flag get_flag reset_flag);

sub import
{
        my @args = grep { $_ ne ':noX11' } @_;
        my $test_runs_without_x11 = @args != @_;
        __PACKAGE__->export_to_level(1,@args);

        # i'm unsure about the line below, needs experimenting
        skip_all unless $test_runs_without_x11 or not $noX11; 
}
```

after this, the code all tests share (i.e.
```perl
if( $Prima::Test::noX11 ) {
    plan skip_all => "Skipping all because noX11";
}
```
) would be going away

4) The component.t. failure went away after the patch in 2).

Hope this helps!

**Comment by Sweet-kid** [comment]
Thanks for answering my questions.

1) For this:
 Unicode.t sets the test environment as if there's just one test, named "support", and skips it. For the rewrite we want a skip_all. The same is with Codec.t - it fakes the test env as if it had 3 tests, passes 2 of them, skips the 3rd. Again, the solution is to skip the rest of the tests.

I've made changes, please check and let me know if it is fine or not.

2) This solved the problem:
    This is partially due to me not thinking that set_flag could be called from object context. This patch to set_flag() will help:

3) For this:
 Sorry i wasn't clear about what is :noX11. This is an import pragma, which should be explicitly tested for in Prima::Test::import() . It's also a bit tricky because it needs to call the Exporter::import(), to have it successfully export set_flag etc names mentioned in @EXPORT. It should work like this:

I've made some extra changes apart from what you said, I checked Exporter's man page and copied some code from there, please let me know if it's fine or not.

4) Yep, it went away with 2) 
The component.t. failure went away after the patch in 2).


Thanks.

**Comment by Sweet-kid** [comment]
travis also passed, yay! (I'm hopeful) :)

**Comment by dk** [comment]
Yes, everything except #3 is ok. With #3 though i'd prefer if you used the logic for import() i've drafted above. noX11 instead of :noX11 is ok, but pushing explicit names to @args is wrong, those needs to be kept intact except the noX11 symbol.

On the other note, there are small things again:

1) done_testing is redundant where there is an explicit plan defined, I think it should be removed there. 

2) image/codec.t doesn't know in advance how many tests it has, because Prima can be compiled with varying number of codecs

3) There are lots of Prima::Test::wait instead of just wait (and also set_flag etc)

4) use strict/use warnings should be in the very beginning of the file, i.e. first modules used. That's almost irrelevant for the end result, but is prettier and that is a common convention.

5) get_flag() etc functions without arguments should be called as get_flag (i.e. no parentheses)

6) "make test" should be working

**Comment by Sweet-kid** [comment]
By "but pushing explicit names to @args is wrong", do you mean this?
+ push @args, qw(create_window set_flag get_flag reset_flag);

2) image/codec.t doesn't know in advance how many tests it has, because Prima can be compiled with varying number of codecs

So I should not declare any plan for Image/Codec.t, right?

**Comment by dk** [comment]
1) yes, this. Consider the case when it will be used as "use Prima::Test ();" which means no imports at all.

2) yes, as long it works well together with skips

**Comment by Sweet-kid** [comment]
for 1) does this make sense https://gist.github.com/Sweet-kid/d275405622c9de0a17a7 ? I totally misunderstood the usage of import!

**Comment by Sweet-kid** [comment]
@dk I've pushed some commits which are solving above issues, please check again and let me know if it's fine or not! Thanks.

**Comment by Sweet-kid** [comment]
travis failed! I'm looking into it, please check later on!

**Comment by Sweet-kid** [comment]
I couldn't reproduce that error on my system. It's saying that I tried to plan twice, which I did, but it's working fine on my system. I'm doing it with prove.

**Comment by dk** [comment]
This is definitely better! Now what I have: 

1) it seems that Image/Stream.t is also, like Image/Codec.t works on variable number of tests. let's revert it back to done_testing

2) what's the purpose of "use Prima::Test ();" in Image/Codec.t? It should be "use Prima::Test (noX11);"

3) likewise, in Image/Colormap.t, it also should be "use Prima::Test (noX11);", along with Object/Object.t, Image/Stream.t, and Image/Import.t

4) other "normal" tests should just use "use Prima::Test;"

you can simulate no-x11 env in your linux setup by running "DISPLAY= make test". Travis tests are run like that I believe.

**Comment by Sweet-kid** [comment]
1) it seems that Image/Stream.t is also, like Image/Codec.t works on variable number of tests. let's revert it back to done_testing

Reverted


2) what's the purpose of "use Prima::Test ();" in Image/Codec.t? It should be "use Prima::Test (noX11);"

It was a typo :(. Fixed now.


3) likewise, in Image/Colormap.t, it also should be "use Prima::Test (noX11);", along with Object/Object.t, Image/Stream.t, and Image/Import.t

Now, I'm confused about this.
 plan skip_all => "skipping all because noX11"
        if( $test_runs_without_x11 && $noX11 );
}
As per my understanding of the above snippet, it means skip all if noX11 is imported (only those tests will import noX11, which cannot be run in noX11 environment) && running in noX11

What I understood from your above comment  is this:
 plan skip_all => "skipping all because noX11"
        if(!$test_runs_without_x11 && $noX11);
}

**Comment by dk** [comment]
1,2) okay, cool

3) well that wasn't I wrote. The original line was "skip_all unless $test_runs_without_x11 or not $noX11" which can be rewritten as "!$test_runs_without_x11 && $noX11" but not as "if( $test_runs_without_x11 && $noX11 )". There seems to be a typo.

Also, the understanding I think is a bit off. Prima is a GUI toolkit, and needs X11. So by default all tests will have it, or skip_all when it is unavailable. Except the minority that doesn't need it, and flags about that with "use Prima::Test qw(noX11)". This minority can run no matter the X11 availability.

Hope this helps

**Comment by Sweet-kid** [comment]
Sorry, I got confused. I've pushed the changes & waiting for travis.

**Comment by Sweet-kid** [comment]
Travis failed with the same errors as before & I couldn't reproduce those errors on my system!

**Comment by dk** [comment]
well I can reproduce those error if I run tests with DISPLAY unset

**Comment by Sweet-kid** [comment]
Hi @dk 

Sorry, I was again doing a stupidity. But now, I've found the problem & pushed the fix. Waiting for travis. Hope it will pass this time.

Thanks.

**Comment by Sweet-kid** [comment]
Yay, travis passed!

**Comment by dk** [comment]
Hi Upasana,

Thank you very much for your contribution! I owe you a drink on the next YAPC :)

Sincerely,
Dmitry

**Comment by Sweet-kid** [comment]
Hi @dk 

Thanks to *you* for all the code reviews! :) I learnt a lot, really helpful! See you at YAPC::EU (hopefully)
